numpy/core/src/npysort/x86-qsort.dispatch.c.src

```diff
@@ -0,0 +1,585 @@
+    COEX_YMM_@vsuf1@(v[0], v[2]); COEX_YMM_@vsuf1@(v[1], v[3]);
+    COEX_YMM_@vsuf1@(v[4], v[6]); COEX_YMM_@vsuf1@(v[5], v[7]);
+    COEX_YMM_@vsuf1@(v[0], v[4]); COEX_YMM_@vsuf1@(v[1], v[2]);
+    COEX_YMM_@vsuf1@(v[5], v[6]); COEX_YMM_@vsuf1@(v[3], v[7]);
+    COEX_YMM_@vsuf1@(v[1], v[5]); COEX_YMM_@vsuf1@(v[2], v[6]);
+    COEX_YMM_@vsuf1@(v[3], v[5]); COEX_YMM_@vsuf1@(v[2], v[4]);
+    COEX_YMM_@vsuf1@(v[3], v[4]);
+    COEX_YMM_@vsuf1@(v[3], v[8]);
+    COEX_YMM_@vsuf1@(v[4], v[8]);
+
+    // technically v[4] needs to be sorted before we pick the correct median,
+    // picking the 4th element works just as well for performance
+    @type_t@* temp = (@type_t@*) &v[4];
+
+    return temp[4];
+}
+
+/*
+ * Parition one ZMM register based on the pivot and returns the index of the
+ * last element that is less than equal to the pivot.
+ */
+static NPY_INLINE NPY_GCC_OPT_3
+npy_int partition_vec_@vsuf1@(@type_t@* arr, npy_intp left, npy_intp right,
+                                const @zmm_t@ curr_vec, const @zmm_t@ pivot_vec,
+                                @zmm_t@* smallest_vec, @zmm_t@* biggest_vec)
+{
+    /* which elements are larger than the pivot */
+    __mmask16 gt_mask = _mm512_cmp_@vsuf1@_mask(curr_vec, pivot_vec, @CMP_GE_OP@);
+    npy_int amount_gt_pivot = _mm_popcnt_u32((npy_int)gt_mask);
+    _mm512_mask_compressstoreu_@vsuf2@(arr + left, _knot_mask16(gt_mask), curr_vec);
+    _mm512_mask_compressstoreu_@vsuf2@(arr + right - amount_gt_pivot, gt_mask, curr_vec);
+    *smallest_vec = _mm512_min_@vsuf1@(curr_vec, *smallest_vec);
+    *biggest_vec = _mm512_max_@vsuf1@(curr_vec, *biggest_vec);
+    return amount_gt_pivot;
+}
+
+/*
+ * Parition an array based on the pivot and returns the index of the
+ * last element that is less than equal to the pivot.
+ */
+static NPY_INLINE NPY_GCC_OPT_3
+npy_intp partition_avx512_@vsuf1@(@type_t@* arr, npy_intp left, npy_intp right,
+                                    @type_t@ pivot, @type_t@* smallest, @type_t@* biggest)
+{
+    /* make array length divisible by 16 , shortening the array */
+    for (npy_int i = (right - left) % 16; i > 0; --i) {
+        *smallest = MIN(*smallest, arr[left]);
+        *biggest = MAX(*biggest, arr[left]);
+        if (arr[left] > pivot) {
+            swap_@TYPE@(arr, left, --right);
+        }
+        else {
+            ++left;
+        }
+    }
+
+    if(left == right)
+      return left; /* less than 16 elements in the array */
+
+    @zmm_t@ pivot_vec = _mm512_set1_@vsuf2@(pivot);
+    @zmm_t@ min_vec = _mm512_set1_@vsuf2@(*smallest);
+    @zmm_t@ max_vec = _mm512_set1_@vsuf2@(*biggest);
+
+    if(right - left == 16) {
+        @zmm_t@ vec = _mm512_loadu_@vsuf3@(arr + left);
+        npy_int amount_gt_pivot = partition_vec_@vsuf1@(arr, left, left + 16, vec, pivot_vec, &min_vec, &max_vec);
+        *smallest = _mm512_reduce_min_@vsuf1@(min_vec);
+        *biggest = _mm512_reduce_max_@vsuf1@(max_vec);
```

Comment:
```suggestion
#ifdef NPY_HAVE_AVX512F_REDUCE
        *smallest = _mm512_reduce_min_@vsuf1@(min_vec);
        *biggest = _mm512_reduce_max_@vsuf1@(max_vec);
#else
        // emulate 
#endif
```
reduce intrinsics aren't backed by native support and not emulated by all compilers that support , see

https://github.com/numpy/numpy/blob/c65bc212ec1987caefba0ea7efe6a55803318de9/numpy/distutils/checks/extra_avx512f_reduce.c#L2-L7

Reply:
Done. Added it to SIMD module.